Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs

```diff
@@ -38,26 +40,49 @@ private void OnEnable()
                     None = new ComponentType[] { typeof(NavPlanning), typeof(NavJumping) }
                 });
             navGroundingSystem = World.DefaultGameObjectInjectionWorld.GetOrCreateSystem<NavGroundingSystem>();
+            debugSystems = World.DefaultGameObjectInjectionWorld.GetOrCreateSystem<DebugSystems>();
+            debugSystems.DrawPathOffset = offsetForDrawPath;
+        }
+
+        private void Update()
+        {
+            navGroundingSystem.IsDebugging = isDebugging && drawUnitVectors;
+            debugSystems.LogNavError = isDebugging && logNavError;
+            debugSystems.DrawPath = isDebugging && drawnPath;
         }
+    }
 
-        private void OnGUI()
+    public class DebugSystems : SystemBase
+    {
+        public bool LogNavError = false;
+        public bool DrawPath = false;
+        public float3 DrawPathOffset = default;
+
+        protected override void OnUpdate()
         {
-            navGroundingSystem.IsDebugging = IsDebugging = isDebugging;
-            navGroundingSystem.DrawUnitVectors = drawUnitVectors;
-            if (!isDebugging || !drawnPath) return;
+            if (LogNavError)
+            {
+                Entities.ForEach((Entity entity, in NavHasProblem navHasProblem) =>
+                {
+                    Debug.Log(string.Format("{0} has a nav problem! {1}", entity, navHasProblem.Value));
+                }).Run();
+            }
 
-            var entityArray = entityQuery.ToEntityArray(Allocator.TempJob);
-            foreach (Entity entity in entityArray)
+            if (DrawPath)
             {
-                var pathBuffer = entityManager.GetBuffer<NavPathBufferElement>(entity);
-                for (int i = 0; i < pathBuffer.Length - 1; ++i)
+                float3 drawPathOffset = DrawPathOffset;
+                Entities.ForEach((Entity entity, in DynamicBuffer<NavPathBufferElement> pathBuffer) =>
                 {
-                    var node = pathBuffer[i];
 
-                    Debug.DrawLine(node.Value + offset, pathBuffer[i + 1].Value + offset, Color.red);
-                }
+                    for (int i = 0; i < pathBuffer.Length - 1; ++i)
+                    {
+                        var node = pathBuffer[i];
+
+                        Debug.DrawLine(node.Value + drawPathOffset, pathBuffer[i + 1].Value + drawPathOffset, Color.red);
```

Comment:
Didn't realise I could use Debug.DrawLine from a System as well now, so moved this into a SystemBase Entities loop.